Doc/library/functools.rst

```diff
@@ -208,6 +208,55 @@ The :mod:`functools` module defines the following functions:
       Returning NotImplemented from the underlying comparison function for
       unrecognised types is now supported.
 
+.. function:: topsort(pairiter)
+   Topologically sort a list of (parent, child) pairs. Return a sequence of
+   the elements in dependency order (parent to child order).
+
+   If a dependency cycle is found it raises :exc:`CycleError`.
+
+   This function can be used directly with a graph given by 2-tuples
+   specifiying the relationships between elements:
+
+      >>> topsort([(1,2), (3,4), (5,6), (1,3), (1,5), (1,6), (2,5)])
+      [1, 2, 3, 5, 4, 6]
+      >>> topsort([(1,2), (1,3), (2,4), (3,4), (5,6), (4,5)])
+      [1, 2, 3, 4, 5, 6]
+
+   If we provide a cyclic graph, :exc:`CycleError` will be raised:
```

Comment:
Perhaps unsurprisingly, I prefer the API used by https://pypi.org/project/toposort/:
```
>>> list(toposort({2: {11},
...                9: {11, 8, 10},
...                10: {11, 3},
...                11: {7, 5},
...                8: {7, 3},
...               }))
[{3, 5, 7}, {8, 11}, {2, 10}, {9}]
```
That is, don't lose the information that 3, 5, and 7 can be used in any order and still satisfy the conditions. There's also a convenience function `toposort_flatten` that returns a single flattened list. Although I'm not wild about the way it sorts the elements in the sets.

Reply:
I think both API's do not overlap completely, as flattening this version will depend on the order of the elements in the set as you mention, while the result in the current one is always deterministic and I do not think you can reconstruct this API with the current one. Maybe we could expose both APIs. Thoughs?

Reply:
Instead of taking a *dict* mapping vertices to *sets* of their out-neighbours, I suggest taking a *mapping* from vertices to *iterables* of their out-neighbours. Then someone who needs a deterministic result can use lists instead of sets for the iterables. (Or ordered sets, or whatever is appropriate for their problem.)